### ISSUE TITLE
dns_get_record Error

### ISSUE THREAD

**Issue by sai-mike** [blind-coder/rcmcarddav#220]
I had RCMCardDAV 2.0.x running correctly on my Roundcube v1.3.6 server.  I dropped the old tables this morning and deleted carddav from roundcube/plugins in order to upgrade.  After using composer to install the new 3.0 version, I logged back in to RC which properly created the 6 RCMCardDAV tables in my database.  When I attempt to recreate my old address book, I get a message that it was successfully created, but it is empty.  Looking in roundcube/logs I find the following errors:

[02-May-2018 12:50:20 America/Los_Angeles] PHP Warning:  dns_get_record(): An unexpected server failure occurred. in /var/www/roundcube/plugins/carddav/carddav_discovery.php on line 252
[02-May-2018 12:50:20 America/Los_Angeles] PHP Warning:  Invalid argument supplied for foreach() in /var/www/roundcube/plugins/carddav/carddav_discovery.php on line 253

The URL I'm using is correct - it worked with v2.0 and I can open it in a browser and retrieve the CardDAV records.

**Comment by blind-coder** [comment]
That should only happen when you do not have the proper SRV/TXT dns records for your addressbooks, but it's not a critical error. Addressbook discovery should continue with well-known URIs. Can you enable DEBUG and DEBUG_HTTP in carddav_common.php, recreate the addressbook and provide the roundcube/logs/carddav* files?

**Comment by sai-mike** [comment]
Here are the files as requested.

[carddav.txt](https://github.com/blind-coder/rcmcarddav/files/1971899/carddav.txt)
[errors.txt](https://github.com/blind-coder/rcmcarddav/files/1971900/errors.txt)

**Comment by blind-coder** [comment]
According to that, an addressbook has been found at https://synology-415.sierraai.com:8443/addressbooks/users/DOMAIN%5Cmike/addressbook, but with an empty name. Have you entered a name for the addressbook during discovery?

**Comment by sai-mike** [comment]
I'm not sure I understand what you mean by empty name. If I enter the URL above in my browser, I get a username/password dialog:
![cred](https://user-images.githubusercontent.com/35618397/39654230-e6cb1c0e-4fa8-11e8-9bba-e375860337ea.png)

When I enter those, I get a page showing all my contacts on the CardDAV Server:
![reply](https://user-images.githubusercontent.com/35618397/39654290-1c8e1814-4fa9-11e8-9c10-a1a6bc9fd9f3.png)

When I create the new addressbook in the Roundcube settings, I am entering a name for the new adressbook:
![settings](https://user-images.githubusercontent.com/35618397/39654371-56a4534c-4fa9-11e8-8055-f88addfbbb0c.png)

**Comment by sai-mike** [comment]
I think I may see what you were referring to - my CardDAV Server is a Synology NAS. It only supports one address book per user (or principal), and those are pre-created for each account.  If I query the CardDAV Server using the URL above, it returns a list of my contacts and below that it returns a list of properties. The displayname is blank - 

Name                         Value
{DAV:}displayname     (no value)

Is your discovery code requiring an addressbook name that the Synology does not provide?

**Comment by sai-mike** [comment]
Prior to creating a CardDAV addressbook, the only table with records is carddav_migrations. After creating an addressbook, there is one record in carddav_addressbooks:
![db_records](https://user-images.githubusercontent.com/35618397/39665690-dddd7840-504c-11e8-9909-0e3e0aa028a5.png)

![db_abook](https://user-images.githubusercontent.com/35618397/39665694-e5ab2752-504c-11e8-8a52-c027a73fb22b.png)

![migrate](https://user-images.githubusercontent.com/35618397/39665697-ecbb8cee-504c-11e8-984d-47e42a3a4027.png)

**Comment by blind-coder** [comment]
That sounds right. So is the addressbook usable now?

**Comment by blind-coder** [comment]
That's weird. Can you get me the same logfiles as before? Set the sync interval to 0, open the addressbook, then check logs/carddav*

**Comment by sai-mike** [comment]
I just opened Roundcube on my desktop so I could gather the info for you and there was a long delay opening it after logging in.  When it opened, my SynCardDAV contacts were there. In roundcube/logs this error was logged in carddav:

[05-May-2018 10:17:49 -0700]: <fmadeq64> An error (status 401) occured while retrieving vcards for addressbook 6. Synchronization aborted.

Here are the errors from the other logs:
[errors.txt](https://github.com/blind-coder/rcmcarddav/files/1976970/errors.txt)
[carddav.warn.txt](https://github.com/blind-coder/rcmcarddav/files/1976971/carddav.warn.txt)

Unfortunately, DEBUG was off when these errors were logged (I didn't know how verbose it was so I turned it off after gathering the first set of logs above).

Even though there are errors logged in carddav.warn about .VCFs that don't exist, I have all 166 Contacts from my CardDAV server in Roundcube.

**Comment by sai-mike** [comment]
I set the sync interval to zero and did some edits, deletions, etc. in my CardDAV addressbook and there are no more errors logged in the roundcube/logs.

**Comment by blind-coder** [comment]
The 401 error is okay, it's an indicator that a less efficient sync mechanism is going to be used.
I'll be going out on a limb here and asume that the error happened because you didn't provide a name for the addressbook. I  need to figure out a way to give proper feedback for that...

As for what caused the initial sync of the entries to fail, I'll look into that.

**Comment by newlukai** [comment]
I'm experiencing the same error. I followed your advice and enabled debug logging. Please find the logs here:

[carddav.log](https://github.com/blind-coder/rcmcarddav/files/2592197/carddav.log)
[errors.log](https://github.com/blind-coder/rcmcarddav/files/2592198/errors.log)

I recently updated my Roundcube installation (hosted by a webhoster) and migrated rcmcarddav from a pretty old version. Before this update I had no problem connecting the CardDAV server. During the migration, there were some issues with the database, but I was able to solve them.
I run my Nextcloud installation on a Raspi at home and provide it via nginx. I use a DynDNS to connect the Nextcloud from outside.
Regarding this issue, nginx returns a 504.

**Comment by hudecof** [comment]
my error is
```
[11-Jul-2019 06:30:56 Europe/Berlin] PHP Warning:  dns_get_record(): An unexpected server failure occurred. in /srv/web/mail/htdocs/roundcubemail/plugins/carddav/carddav_discovery.php on line 252
[11-Jul-2019 06:30:56 Europe/Berlin] PHP Warning:  Invalid argument supplied for foreach() in /srv/web/mail/htdocs/roundcubemail/plugins/carddav/carddav_discovery.php on line 253
```

Code looks like
```
        private function find_baseurls($services)
        {{{
        foreach($services as &$service) {
                $baseurl = $service['baseurl'];
                $dnssrv  = $service['dnssrv'];

                $paths = array();
                rcmail::write_log('carddavX', $services);
                $dnsresults = dns_get_record($dnssrv, DNS_TXT);
                foreach($dnsresults as $dnsresult) {
                        if($dnsresult['host'] != $dnssrv) continue;
```

and the input is
```
[11-Jul-2019 06:30:56 +0200]: <fgd4kafc> array (
  0 =>
  array (
    'host' => 'cloud.apps.hudecof.net',
    'port' => 443,
    'baseurl' => 'https://cloud.apps.hudecof.net:443',
  ),
)
```

so there is no ** dnssrv** key in the dictionary.

**Comment by spyco2671** [comment]
I have the same problem with @hudecof

**Comment by hudecof** [comment]
there are atleast 2 places in the code, where is not dict key checking. I have fixed thats parts, coudl submit PR.
@spyco2671 try to enable debug mode, that helped me a lot to find out what's wrong.

**Comment by hudecof** [comment]
my root problem was 2FA in nextcloud. I wanted to import default carddav `$prefs['Personal'] = array(` As the 2FA is enabled I could not use IMAP password.

**Comment by spyco2671** [comment]
@hudecof enabling debugging causes the same problem as you do, but in my nextcloud I did not qualify the 2fa

**Comment by hudecof** [comment]
In both cases I added dict key checking, try t find this code and add `array_key_exists ` asI did.

##  carddav_discovery.php
```
        private function find_baseurls($services)
        {{{
        foreach($services as &$service) {
                if (!array_key_exists('dnssrv', $service)) {
                        continue;
                }
                $baseurl = $service['baseurl'];
                $dnssrv  = $service['dnssrv'];
```

## carddav_discovery.php lines about 86
```
        // now check each of them until we find something (or don't)
        foreach($services as $service) {
                $cdfopen_cfg['url'] = $service['baseurl'];
                if (!array_key_exists('paths', $service)) {
                        continue;
                }
                foreach($service['paths'] as $path) {
                        $aBooks = $this->retrieve_addressbooks($path, $cdfopen_cfg);
                        if(is_array($aBooks) && count($aBooks)>0)
                                return $aBooks;
                }
        }
```

**Comment by ledufakademy** [comment]
Same issue here, with rc 1.4 RC1 , and last release of Carddav : in log searching https ... while i put http in preset ?

**Comment by mstilkerich** [comment]
The code above (which was merged with commit e6138e23a31ec099306383c42a9a02231d8ef2b1 to master in the meantime) breaks discovery for services without DNS SRV records set up. The v4 development branch has proper fixes for the discovery, so if you are affected by this issue please try that branch.

Instructions here:
https://github.com/blind-coder/rcmcarddav/tree/v4
